File: Rust/gameOfLife.rs
use std::collections::HashSet;
use std::io::Write;
use std::io;
use std::thread;
use std::time::Duration;


type Cell = (i32, i32);
type Grid = HashSet<Cell>;


fn neighbors (x: i32, y: i32) -> Vec<Cell> {
    vec![
        (x-1, y-1), (x, y-1), (x+1, y-1),
        (x-1, y),             (x+1, y),
        (x-1, y+1), (x, y+1), (x+1, y+1),
    ]
}


fn count_live_neighbors(grid: &Grid, cell: &Cell) -> usize {
    neighbors(cell.0, cell.1)
        .iter()
        .filter(|n| grid.contains(n))
        .count()
}


fn should_live(grid: &Grid, cell: &Cell) -> bool {
    let live_neighbors = count_live_neighbors(grid, cell);
    let is_alive = grid.contains(cell);
    
    match (is_alive, live_neighbors) {
        (true, 2) | (true, 3)  => true,
        (false, 3) => true,
        _ => false

    }
}

fn cells_to_check(grid: &Grid) -> HashSet<Cell> {
    let mut check = grid.clone();

    for &cell in grid.iter(){
        for neigh in neighbors(cell.0, cell.1) {
            check.insert(neigh);
        }
    }
    check
}

fn step(grid: &Grid) -> Grid {
    cells_to_check(grid)
        .iter()
        .filter(|c| should_live(grid, c))
        .copied()
        .collect()
}

fn display_grid(width: i32, height: i32, grid: &Grid) -> String {
    let mut out = String::new();

    for y in 0..height {
        for x in 0..width {
            if grid.contains(&(x, y)) {
                out.push('#')
            }
            else {
                out.push('.')
            }
        }
        out.push('\n')
    }
    out
}


fn clear_screen() {
    print!("\x1B[2J\x1B[H");
    io::stdout().flush().unwrap();
}


fn run_game(width: i32, height: i32, generations: usize, delay_ms: u64, grid: &Grid) {
    let mut current_grid = grid.clone();


    for gen in 0..generations {
        clear_screen();
        println!("Generation: {} | Population: {}", gen, current_grid.len());
        println!("{}", display_grid(width, height, &current_grid));
        io::stdout().flush().unwrap();

        thread::sleep(Duration::from_millis(delay_ms));
        current_grid = step(&current_grid);
    }
}


fn run_game_infinite(width: i32, height: i32, delay_ms: u64, grid: &Grid) {
    let mut current_grid = grid.clone();
    let mut gen = 0;


    loop {
        clear_screen();
        println!("Generation: {} | Population: {}", gen, current_grid.len());
        println!("{}", display_grid(width, height, &current_grid));
        io::stdout().flush().unwrap();

        thread::sleep(Duration::from_millis(delay_ms));
        current_grid = step(&current_grid);
        gen += 1;

    }
}


fn glider(x: i32, y: i32) -> Grid {
    vec![
        (x+1, y), (x+2, y+1), (x, y+2), (x+1, y+2), (x+2, y+2)
    ].into_iter().collect()
}

fn blinker(x: i32, y: i32) -> Grid {
    vec![(x, y), (x+1, y), (x+2, y)].into_iter().collect()
}

fn block(x: i32, y: i32) -> Grid {
    vec![(x, y), (x+1, y), (x, y+1), (x+1, y+1)].into_iter().collect()
}

fn toad(x: i32, y: i32) -> Grid {
    vec![
        (x+1, y), (x+2, y), (x+3, y),
        (x, y+1), (x+1, y+1), (x+2, y+1)
    ].into_iter().collect()
}

fn r_pentomino(x: i32, y: i32) -> Grid {
    vec![
        (x+1, y), (x+2, y),
        (x, y+1), (x+1, y+1),
        (x+1, y+2)
    ].into_iter().collect()
}

fn gosper_glider_gun(x: i32, y: i32) -> Grid {
    let mut cells = Vec::new();
    
    cells.extend(vec![(x, y+4), (x, y+5), (x+1, y+4), (x+1, y+5)]);
    
    cells.extend(vec![
        (x+10, y+4), (x+10, y+5), (x+10, y+6),
        (x+11, y+3), (x+11, y+7),
        (x+12, y+2), (x+12, y+8),
        (x+13, y+2), (x+13, y+8),
        (x+14, y+5),
        (x+15, y+3), (x+15, y+7),
        (x+16, y+4), (x+16, y+5), (x+16, y+6),
        (x+17, y+5),
    ]);
    
    cells.extend(vec![
        (x+20, y+2), (x+20, y+3), (x+20, y+4),
        (x+21, y+2), (x+21, y+3), (x+21, y+4),
        (x+22, y+1), (x+22, y+5),
        (x+24, y), (x+24, y+1), (x+24, y+5), (x+24, y+6),
    ]);
    
    cells.extend(vec![(x+34, y+2), (x+34, y+3), (x+35, y+2), (x+35, y+3)]);
    
    cells.into_iter().collect()
}


fn get_user_input(prompt: &str) -> String {
    println!("{}", prompt);
    io::stdout().flush().unwrap();

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn main() {
    println!("Conway's Game of Life");
    println!("====================");
    println!();
    println!("Choose a demo:");
    println!("1. Glider (moves across screen)");
    println!("2. Oscillators (blinker and toad)");
    println!("3. R-pentomino (chaotic growth)");
    println!("4. Glider Gun (creates gliders) - INFINITE");
    println!("5. Mixed patterns");
    println!();
    println!("Run mode:");
    println!("a. Limited generations (stops automatically)");
    println!("b. Infinite mode (run forever, press Ctrl+C to stop)");
    println!();



    let pattern_choice = get_user_input("Enter the pattern choice (1-5)");
    let mode_choice = get_user_input("Enter mode (a/b): ");
    let speed_input = get_user_input("Enter speed in ms (50-1000, default 200): ");


    let speed = if speed_input.is_empty() {
        200
    } else {
        speed_input.parse::<u64>().unwrap_or(200).clamp(50, 1000)
    };

    let is_infinite = mode_choice == "b";

    println!("\nStarting simulation...\n");

    if is_infinite {
        println!("INFINITE MODE - Press Ctrl+C to stop");
    } else {
        println!("LIMITED MODE - Will stop automatically");
    }

    println!("Speed: {}ms per generation", speed);
    thread::sleep(Duration::from_secs(1));


    match pattern_choice.as_str() {
        "1" => {
            let pattern = glider(5, 5);
            if is_infinite {
                run_game_infinite(50, 25, speed, &pattern);
            } else {
                run_game(50, 25, 100, speed, &pattern);
            }
        },
        "2" => {
            let mut pattern = blinker(10, 10);
            pattern.extend(toad(20, 10));
            if is_infinite {
                run_game_infinite(40, 20, speed, &pattern);
            } else {
                run_game(40, 20, 50, speed, &pattern);
            }
        },
        "3" => {
            let pattern = r_pentomino(30, 15);
            if is_infinite {
                run_game_infinite(60, 30, speed, &pattern);
            } else {
                run_game(60, 30, 200, speed, &pattern);
            }
        },
        "4" => {
            let pattern = gosper_glider_gun(5, 10);
            if is_infinite {
                run_game_infinite(80, 40, speed, &pattern);
            } else {
                run_game(80, 40, 300, speed, &pattern);
            }
        },
        "5" => {
            let mut pattern = glider(5, 5);
            pattern.extend(blinker(25, 12));
            pattern.extend(block(40, 18));
            pattern.extend(toad(15, 20));
            if is_infinite {
                run_game_infinite(50, 25, speed, &pattern);
            } else {
                run_game(50, 25, 150, speed, &pattern);
            }
        },
        _ => {
            println!("Invalid choice. Running glider demo...");
            let pattern = glider(5, 5);
            if is_infinite {
                run_game_infinite(50, 25, speed, &pattern);
            } else {
                run_game(50, 25, 100, speed, &pattern);
            }
        }
    }

    println!("\nSimulation complete!");
}

File: Go/gameOfLife.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type Cell struct {
	X, Y int
}

type Grid map[Cell]bool

type PatternConfig struct {
	Pattern     Grid
	Width       int
	Height      int
	Generations int
}

func neighbors(x, y int) []Cell {
	return []Cell{
		{x - 1, y - 1}, {x, y - 1}, {x + 1, y - 1},
		{x - 1, y}, {x + 1, y},
		{x - 1, y + 1}, {x, y + 1}, {x + 1, y + 1},
	}
}

func countLiveNeighbors(grid Grid, cell Cell) int {
	count := 0
	for _, n := range neighbors(cell.X, cell.Y) {
		if grid[n] {
			count++
		}
	}
	return count
}

func shouldLive(grid Grid, cell Cell) bool {
	liveNeighbors := countLiveNeighbors(grid, cell)
	isAlive := grid[cell]
	if isAlive && (liveNeighbors == 2 || liveNeighbors == 3) {
		return true
	} else if !isAlive && liveNeighbors == 3 {
		return true
	} else {
		return false
	}
}

func cellsToCheck(grid Grid) Grid {
	check := make(Grid)
	for cell := range grid {
		check[cell] = true
		for _, n := range neighbors(cell.X, cell.Y) {
			check[n] = true
		}
	}
	return check
}

func step(grid Grid) Grid {
	newGrid := make(Grid)
	for cell := range cellsToCheck(grid) {
		if shouldLive(grid, cell) {
			newGrid[cell] = true
		}
	}
	return newGrid
}

func displayGrid(width, height int, grid Grid) string {
	var sb strings.Builder
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			if grid[Cell{x, y}] {
				sb.WriteString("#")
			} else {
				sb.WriteString(".")
			}
		}
		sb.WriteString("\n")
	}
	return sb.String()
}

func clearScreen() {
	var cmd *exec.Cmd
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", "cls")
	} else {
		cmd = exec.Command("clear")
	}
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func runGame(width, height, generations, delayMs int, grid Grid) {
	currentGrid := copyGrid(grid)
	for gen := 0; gen < generations; gen++ {
		clearScreen()
		fmt.Printf("Generation: %d | Population: %d\n", gen, len(currentGrid))
		fmt.Print(displayGrid(width, height, currentGrid))
		time.Sleep(time.Duration(delayMs) * time.Millisecond)
		currentGrid = step(currentGrid)
	}
}

func runGameInfinite(width, height, delayMs int, grid Grid) {
	currentGrid := copyGrid(grid)
	gen := 0
	for {
		clearScreen()
		fmt.Printf("Generation: %d | Population: %d\n", gen, len(currentGrid))
		fmt.Print(displayGrid(width, height, currentGrid))
		fmt.Println("\n[Press Ctrl+C to stop]")
		time.Sleep(time.Duration(delayMs) * time.Millisecond)
		currentGrid = step(currentGrid)
		gen++
	}
}

func copyGrid(grid Grid) Grid {
	newGrid := make(Grid)
	for cell := range grid {
		newGrid[cell] = true
	}
	return newGrid
}

func glider(x, y int) Grid {
	return Grid{
		{x + 1, y}:     true,
		{x + 2, y + 1}: true,
		{x, y + 2}:     true,
		{x + 1, y + 2}: true,
		{x + 2, y + 2}: true,
	}
}

func blinker(x, y int) Grid {
	return Grid{
		{x, y}:     true,
		{x + 1, y}: true,
		{x + 2, y}: true,
	}
}

func block(x, y int) Grid {
	return Grid{
		{x, y}:         true,
		{x + 1, y}:     true,
		{x, y + 1}:     true,
		{x + 1, y + 1}: true,
	}
}

func toad(x, y int) Grid {
	return Grid{
		{x + 1, y}:     true,
		{x + 2, y}:     true,
		{x + 3, y}:     true,
		{x, y + 1}:     true,
		{x + 1, y + 1}: true,
		{x + 2, y + 1}: true,
	}
}

func rPentomino(x, y int) Grid {
	return Grid{
		{x + 1, y}:     true,
		{x + 2, y}:     true,
		{x, y + 1}:     true,
		{x + 1, y + 1}: true,
		{x + 1, y + 2}: true,
	}
}

func gosperGliderGun(x, y int) Grid {
	grid := make(Grid)
	cells := []Cell{
		{x, y + 4}, {x, y + 5}, {x + 1, y + 4}, {x + 1, y + 5},
	}
	cells = append(cells,
		Cell{x + 10, y + 4}, Cell{x + 10, y + 5}, Cell{x + 10, y + 6},
		Cell{x + 11, y + 3}, Cell{x + 11, y + 7},
		Cell{x + 12, y + 2}, Cell{x + 12, y + 8},
		Cell{x + 13, y + 2}, Cell{x + 13, y + 8},
		Cell{x + 14, y + 5},
		Cell{x + 15, y + 3}, Cell{x + 15, y + 7},
		Cell{x + 16, y + 4}, Cell{x + 16, y + 5}, Cell{x + 16, y + 6},
		Cell{x + 17, y + 5},
	)
	cells = append(cells,
		Cell{x + 20, y + 2}, Cell{x + 20, y + 3}, Cell{x + 20, y + 4},
		Cell{x + 21, y + 2}, Cell{x + 21, y + 3}, Cell{x + 21, y + 4},
		Cell{x + 22, y + 1}, Cell{x + 22, y + 5},
		Cell{x + 24, y}, Cell{x + 24, y + 1}, Cell{x + 24, y + 5}, Cell{x + 24, y + 6},
	)
	cells = append(cells,
		Cell{x + 34, y + 2}, Cell{x + 34, y + 3}, Cell{x + 35, y + 2}, Cell{x + 35, y + 3},
	)
	for _, cell := range cells {
		grid[cell] = true
	}
	return grid
}

func mergeGrids(grids ...Grid) Grid {
	result := make(Grid)
	for _, grid := range grids {
		for cell := range grid {
			result[cell] = true
		}
	}
	return result
}

func getUserInput(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func getSpeedInput(prompt string, defaultVal, minVal, maxVal int) int {
	input := getUserInput(prompt)
	if input == "" {
		return defaultVal
	}
	speed, err := strconv.Atoi(input)
	if err != nil {
		return defaultVal
	}
	if speed < minVal {
		return minVal
	}
	if speed > maxVal {
		return maxVal
	}
	return speed
}

func getPatternConfig(choice string) PatternConfig {
	switch choice {
	case "1":
		return PatternConfig{glider(5, 5), 50, 25, 100}
	case "2":
		return PatternConfig{mergeGrids(blinker(10, 10), toad(20, 10)), 40, 20, 50}
	case "3":
		return PatternConfig{rPentomino(30, 15), 60, 30, 200}
	case "4":
		return PatternConfig{gosperGliderGun(5, 10), 80, 40, 300}
	case "5":
		return PatternConfig{
			mergeGrids(glider(5, 5), blinker(25, 12), block(40, 18), toad(15, 20)),
			50, 25, 150,
		}
	default:
		return PatternConfig{glider(5, 5), 50, 25, 100}
	}
}

func printMenu() {
	fmt.Println("Conway's Game of Life")
	fmt.Println("====================")
	fmt.Println()
	fmt.Println("Choose a demo:")
	fmt.Println("1. Glider (moves across screen)")
	fmt.Println("2. Oscillators (blinker and toad)")
	fmt.Println("3. R-pentomino (chaotic growth)")
	fmt.Println("4. Glider Gun (creates gliders) - INFINITE")
	fmt.Println("5. Mixed patterns")
	fmt.Println()
	fmt.Println("Run mode:")
	fmt.Println("a. Limited generations (stops automatically)")
	fmt.Println("b. Infinite mode (run forever, press Ctrl+C to stop)")
	fmt.Println()
}

func main() {
	printMenu()
	patternChoice := getUserInput("Enter pattern choice (1-5): ")
	modeChoice := getUserInput("Enter mode (a/b): ")
	speed := getSpeedInput("Enter speed in ms (50-1000, default 200): ", 200, 50, 1000)
	isInfinite := strings.ToLower(modeChoice) == "b"
	config := getPatternConfig(patternChoice)
	if patternChoice < "1" || patternChoice > "5" {
		fmt.Println("Invalid choice. Running glider demo...")
	}
	fmt.Println("\nStarting simulation...\n")
	if isInfinite {
		fmt.Println("INFINITE MODE - Press Ctrl+C to stop")
	} else {
		fmt.Println("LIMITED MODE - Will stop automatically")
	}
	fmt.Printf("Speed: %dms per generation\n", speed)
	time.Sleep(1 * time.Second)
	if isInfinite {
		runGameInfinite(config.Width, config.Height, speed, config.Pattern)
	} else {
		runGame(config.Width, config.Height, config.Generations, speed, config.Pattern)
	}
	fmt.Println("\nSimulation complete!")
}
